proposals/low-level-struct-improvements.md

```diff
@@ -50,248 +45,233 @@ readonly ref struct Span<T>
+- Else `ref` fields can only be initialized with known heap values.
 
-- If the constructor contains any `ref`, `out` or `in` parameters, and the arguments do not all refer to the heap, then the *safe-to-escape* of the return will be the current scope
-- Else if the constructor contains any `ref struct` parameters then the *safe-to-escape* of the return will be the current scope
-- Else the *safe-to-escape* will be the outside the method scope
+These rules ensure that the caller can understand when constructors will or will not capture parameters by `ref`. 
 
-Let's examine these rules in the context of samples to better understand their impact.
+Next the rules for method invocation will change as follows when the target method is annotated with `[RefFieldEscapes]`:
+- If the current method is not annotated with `[RefFieldEscapes]` then
+    - If all arguments have a *ref-safe-to-escape* scope of *heap* then the *safe-to-escape* scope of the return is to the *calling method*
+    - Else the *safe-to-escape* scope is to the *current method*
+- Else the *safe-to-escape* scope of the return is the minimum of 
+    - The existing *safe-to-escape* calculation for method invocation
+    - All of the *ref-safe-to-escape* values of `ref`, `in` and `out` arguments
 
-```cs
+The design of `[RefFieldEscape]` will be discussed in detail [later in the proposal](#reffieldescapes).
+
+Let's examine these rules in the context of samples to better understand their impact and how they maintain the required compat considerations.
+
+```c#
 ref struct RS
 {
     ref int _field;
 
     public RS(int[] array, int index)
     {
+        // Okay: even though there is no [RefFieldEscapes] annotation the fields are initialized
+        // with known heap values
         _field = ref array[index];
     }
 
+    [RefFieldEscapes]
     public RS(ref int i)
     {
+        // Okay: the constructor is annotated with [RefFieldEscapes] which allows for fields to be
+        // initialized with any value that is ref-safe-to-escape to calling method. That is the 
+        // case for `i` hence this is legal
         _field = ref i;
     }
 
-    static RS CreateRS(ref int i)
+    public RS(ref (int, int) tuple)
     {
-        // The implementation of this method is irrelevant to the safety rule
-        // examples below. The returned value is always *safe-to-escape* outside
-        // the enclosing method scope
+        // Error: even though the ref-safe-to-escape of `tuple.Item1` is to the calling method, the 
+        // constructor is not annotated with [RefFieldEscapes] hence only known heap values are 
+        // allowed
+        _field = ref tuple.Item1;
     }
 
-    static RS RuleExamples(ref int i, int[] array)
+    static RS CreateRS(ref int parameter)
     {
-        var rs1 = new RS(ref i);
+        // Error: The RS(ref int) ctor is annotated as [RefFieldEscapes] however CreateRS is 
```

Comment:
Generics should not add any particular complication here. This problem only comes into effect when: 

- A `ref struct` with a `ref` field is returned from a method. Either by the method return or via a `out` / `ref` parameter
- There is an additional `ref`, `out` or `in` parameter of any type

Hence generics don't impact this much at all because `ref struct` can't be generic. Even if we add support for generic `ref struct` then it's an opt in scenario, like `where T : ref struct`, hence it should still be very limited. 

Note: in searching through the BCL today I've only found a very limited number of these methods and most are non-public. If you are aware of more I'd like to hear about them.